ebu_tt_live/bindings/validation.py

```diff
@@ -159,43 +194,223 @@ def _post_timing_set_attribute(self, attr_use):
+                    raise SemanticValidationError(
+                        ERR_SEMANTIC_VALIDATION_TIMING_TYPE.format(
+                            attr_type=type(self.begin),
+                            attr_value=self.begin,
+                            attr_name='begin',
+                            time_base=time_base
+                        )
                     )
-                )
-        if hasattr(self, 'dur') and self.dur is not None and hasattr(self.dur, 'compatible_timebases'):
-            timebases = self.dur.compatible_timebases()
-            if time_base not in timebases['dur']:
-                raise SemanticValidationError(
-                    ERR_SEMANTIC_VALIDATION_TIMING_TYPE.format(
-                        attr_type=type(self.dur),
-                        attr_value=self.dur,
-                        attr_name='dur',
-                        time_base=time_base
+        if self.end is not None:
+            if hasattr(self.end, 'compatible_timebases'):
+                # Check typing of end attribute against the timebase
+                timebases = self.end.compatible_timebases()
+                if time_base not in timebases['end']:
+                    raise SemanticValidationError(
+                        ERR_SEMANTIC_VALIDATION_TIMING_TYPE.format(
+                            attr_type=type(self.end),
+                            attr_value=self.end,
+                            attr_name='end',
+                            time_base=time_base
+                        )
                     )
-                )
-        if hasattr(self, 'end') and self.end is not None and hasattr(self.end, 'compatible_timebases'):
-            timebases = self.end.compatible_timebases()
-            if time_base not in timebases['end']:
-                raise SemanticValidationError(
-                    ERR_SEMANTIC_VALIDATION_TIMING_TYPE.format(
-                        attr_type=type(self.end),
-                        attr_value=self.end,
-                        attr_name='end',
-                        time_base=time_base
-                    )
-                )
 
+
+class BodyTimingValidationMixin(TimingValidationMixin):
+    """
+    The body element seems to be exception from too many rules and makes one common validator pretty difficult
+    to manage. This subclass is meant to call all the extensions/limitations for the body element that does not apply
+    to timed containers in general in the EBU-TT-Live spec.
+    """
+
+    def _pre_init_variables(self, dataset, element_content):
+        super(BodyTimingValidationMixin, self)._pre_init_variables(dataset, element_content)
+        self._dur_timedelta = self.dur and self.dur.timedelta or None
+
+    def _post_cleanup_variables(self):
+        del self._dur_timedelta
+        super(BodyTimingValidationMixin, self)._post_cleanup_variables()
+
+    def _pre_calculate_end(self):
+        # This is all for the body element because of the dur attribute
```

Comment:
If all those ifs and elifs are designed to test all the possible combinations of `dur` present and `end|begin` being absent or present the structure is maybe harder to read than it could be. Why not:

```python
        if self._dur_timedelta is not None
            # test all the combinations of begin and end being absent or present, setting proposed_end as we go
            ...
            # assign the value
            self._pre_assign_end(proposed_end)
        else
            # Fallback case if there is no duration specified the same as the other containers
            super(BodyTimingValidationMixin, self)._pre_calculate_end()
```

for example? Just wondering if there are secret cases or magic in here that are not obvious...